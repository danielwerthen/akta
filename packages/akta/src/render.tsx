import { AktaNode } from './types';
import { DependencyMap } from './dependency-map';
import { prepare } from './prepare';
import { renderingContext, RenderingContext } from './rendering-context';

export function render(
  blueprint: AktaNode,
  parent: HTMLElement,
  ctx: RenderingContext = new RenderingContext()
): () => void {
  renderingContext.setContextUnsafe(ctx);
  try {
    const node = prepare(blueprint, new DependencyMap());
    console.log(node);
    if (Array.isArray(node)) {
      parent.append(...node);
    } else {
      parent.append(node);
    }
  } finally {
    renderingContext.resetContextUnsafe(ctx);
  }
  return () => ctx.terminate();
}
