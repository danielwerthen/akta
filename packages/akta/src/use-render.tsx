import { Observable, ReplaySubject } from 'rxjs';
import { dependecyContext } from './dependencies';
import { prepare } from './prepare';
import { renderingContext, RenderingState } from './rendering-context';
import { AktaNode, isAktaElement } from './types';

export function useRender() {
  const deps = dependecyContext.getContext();
  return function render(node: AktaNode): ChildNode | Observable<ChildNode> {
    if (!isAktaElement(node) || typeof node.type !== 'string') {
      throw new Error(
        'UseRender function only accepts a single html tag at the top.'
      );
    }
    const parent = renderingContext.getContext();
    const rendering = parent.spawn();
    renderingContext.setContextUnsafe(rendering);
    try {
      const instance = prepare(node, deps);
      if (Array.isArray(instance)) {
        throw new Error('UseRender output was unexpectedly an array.');
      }
      if (rendering.state.value === RenderingState.active) {
        return instance;
      }
      const obs = new ReplaySubject<ChildNode>(1);
      const sub = rendering.state.subscribe(val => {
        if (val === RenderingState.active) {
          obs.next(instance);
          obs.complete();
        }
      });
      rendering.subscriptions.push(sub);
      return obs;
    } finally {
      renderingContext.resetContextUnsafe(rendering);
    }
  };
}
