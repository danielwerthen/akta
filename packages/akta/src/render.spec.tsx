/** @jsxImportSource . */
import { BehaviorSubject, map, Observable, Subject } from 'rxjs';
import { AktaNode } from '../dist';
import { render } from './render';
import { useRender } from './use-render';

describe('render', () => {
  let parent: HTMLElement;
  let teardown: () => void;
  function doRender(blueprint: AktaNode) {
    teardown = render(blueprint, parent);
  }
  beforeEach(() => {
    parent = document.createElement('div');
  });
  afterEach(() => {
    if (teardown) {
      teardown();
    }
    parent.innerHTML = '';
  });
  it('should rendering akta blueprint', () => {
    doRender(<p>Blueprint</p>);
    expect(parent).toMatchSnapshot();
  });
  it('should use prepared rendering akta blueprint', () => {
    function Component({
      children,
      order,
    }: {
      order: Observable<number[]>;
      children: AktaNode[];
    }) {
      const doRender = useRender();
      const arr = children.map(doRender);
      return <div>{order.pipe(map(idx => idx.map(i => arr[i])))}</div>;
    }
    const subfn = jest.fn();
    subfn.mockReturnValue('sub-fn');
    function SubComponent() {
      return <p>{subfn()}</p>;
    }
    const subj = new BehaviorSubject([0, 1, 2]);
    doRender(
      <Component order={subj}>
        <div>
          Alpha
          <SubComponent />
        </div>
        <div>
          Beta
          <SubComponent />
        </div>
        <div>
          Theta
          <SubComponent />
        </div>
      </Component>
    );
    expect(parent).toMatchSnapshot();
    subj.next([2, 1, 0]);
    expect(parent).toMatchSnapshot();
    subj.next([0, 1, 0, 2, 0]);
    expect(parent).toMatchSnapshot();
    expect(subfn.mock.calls.length).toEqual(3);
  });
  it('should allow async elements in prepared nodes', () => {
    function Component({
      children,
      order,
    }: {
      order: Observable<number[]>;
      children: AktaNode[];
    }) {
      const doRender = useRender();
      const arr = children.map(doRender);
      return <div>{order.pipe(map(idx => idx.map(i => arr[i])))}</div>;
    }
    const subject = new Subject<number>();
    function SubComponent() {
      return <p>{subject}</p>;
    }
    const subj = new BehaviorSubject([1, 1, 0]);
    doRender(
      <Component order={subj}>
        <div>
          Alpha
          <SubComponent />
        </div>
        <div>
          Beta
          <SubComponent />
        </div>
        <div>
          Theta
          <SubComponent />
        </div>
      </Component>
    );
    expect(parent).toMatchSnapshot();
    subject.next(44);
    expect(parent).toMatchSnapshot();
  });
});
