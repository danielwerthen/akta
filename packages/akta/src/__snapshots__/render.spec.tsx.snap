// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should allow async elements in prepared nodes 1`] = `
<div>
  <div>
    
    
    
  </div>
</div>
`;

exports[`render should allow async elements in prepared nodes 2`] = `
<div>
  <div>
    <div>
      Beta
      <p>
        44
      </p>
    </div>
    <div>
      Alpha
      <p>
        44
      </p>
    </div>
  </div>
</div>
`;

exports[`render should rendering akta blueprint 1`] = `
<div>
  <p>
    Blueprint
  </p>
</div>
`;

exports[`render should use prepared rendering akta blueprint 1`] = `
<div>
  <div>
    <div>
      Alpha
      <p>
        sub-fn
      </p>
    </div>
    <div>
      Beta
      <p>
        sub-fn
      </p>
    </div>
    <div>
      Theta
      <p>
        sub-fn
      </p>
    </div>
  </div>
</div>
`;

exports[`render should use prepared rendering akta blueprint 2`] = `
<div>
  <div>
    <div>
      Theta
      <p>
        sub-fn
      </p>
    </div>
    <div>
      Beta
      <p>
        sub-fn
      </p>
    </div>
    <div>
      Alpha
      <p>
        sub-fn
      </p>
    </div>
  </div>
</div>
`;

exports[`render should use prepared rendering akta blueprint 3`] = `
<div>
  <div>
    <div>
      Beta
      <p>
        sub-fn
      </p>
    </div>
    <div>
      Theta
      <p>
        sub-fn
      </p>
    </div>
    <div>
      Alpha
      <p>
        sub-fn
      </p>
    </div>
  </div>
</div>
`;
